src/systems/triggered_publisher/TriggeredPublisher.hh

```diff
@@ -243,8 +291,136 @@ namespace systems
+      {
+        ignmsg << "Service call [" << srvInfo.servName
+               << "] succeeded. Response message [" << std::boolalpha
+               << msg.data() << "]\n";
+      }
+      public: virtual void operator() (const msgs::Empty &msg,
+                                       ServiceOutputInfo &srvInfo)
+      {
+        ignmsg << "Service call [" << srvInfo.servName
+               << "] succeeded. Response message [" << msg.unused() 
+               << "]\n";
+      }
+      public: virtual void operator() (const msgs::Pose &msg,
+                                       ServiceOutputInfo &srvInfo)
+      {
+        ignmsg << "Service call [" << srvInfo.servName
+               << "] succeeded. Response message [" 
+               << "name: " << msg.name() << "pos: {x: "
+               << msg.position().x() << ", y: "
+               << msg.position().y() << ", z: "
+               << msg.position().z() << "}]\n";
+      }
+      //NOTE: add more replys for different types
+    };
+
+    private: template<typename ReplyT>
+    std::function<void(const ReplyT &rep, const bool res)> ReplyCallback(
+      ServiceOutputInfo& serviceInfo)
+    {
+      return std::function<void(const ReplyT &, const bool)>(
+        [&serviceInfo] (const ReplyT &msg, const bool res){
+          if (res)
+          {
+            ReplyHandler()(msg, serviceInfo);
+          }
+          else
+          {
+            ignerr << "Service call [" << serviceInfo.servName
+                   << "] failed\n";
+          }
+        });
+    }
+
+    private: template<typename ReplyT, typename RequestT>
+    bool HandleReply(ServiceOutputInfo& serviceInfo, RequestT& req)
+    {
+      // Non-blocking request with request type and callback
+      auto reqCb = ReplyCallback<ReplyT>(serviceInfo);
+      return this->node.Request(serviceInfo.servName, req, reqCb);
+    }
+
+    private: template<typename ReplyT>
+    bool HandleNoRequestMsg(ServiceOutputInfo& serviceInfo)
+    {
+      auto reqCb = ReplyCallback<ReplyT>(serviceInfo);
+      return this->node.Request(serviceInfo.servName, reqCb);
+    }
+
+    private: template<typename RequestT>
+    bool HandleNoReply(ServiceOutputInfo& serviceInfo, RequestT& req)
+    {
+      // Non-blocking request with no callback
+      return this->node.Request(serviceInfo.servName, req);
+    }
+
+    #define HANDLE_REPLY(repT, typeString) \
+    if (serviceInfo.repType == typeString && \
+        serviceInfo.reqType.empty() && !isProcessing) \
+      { \
+        executed = HandleNoRequestMsg<repT>(serviceInfo); \
+        isProcessing = true; \
+      } \
+    else if (serviceInfo.repType.empty() && !isProcessing) \
+      { \
+        executed = HandleNoReply<reqT>(serviceInfo, *req); \
+        isProcessing = true; \
+      } \
+    else if (serviceInfo.repType == typeString && !isProcessing) \
+      { \
+        executed = HandleReply<repT, reqT>(serviceInfo, *req); \
+        isProcessing = true; \
+      } \
```

Comment:
Let's try to avoid the macro here again.  We should probably be able to do something similar to my suggestion of `handle_request` with a templated function here.